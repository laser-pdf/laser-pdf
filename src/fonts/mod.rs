pub mod builtin;
pub mod truetype;

use std::ops::Range;

pub struct GeneralMetrics {
    pub ascent: u32,
    pub line_height: u32,
}

#[derive(Debug, Clone)]
pub struct ShapedGlyph {
    pub unsafe_to_break: bool,
    pub glyph_id: u32,
    pub text_range: Range<usize>,
    /// without kerning
    pub x_advance_font: i32,
    pub x_advance: i32,
    pub x_offset: i32,
    pub y_advance: i32,
    pub y_offset: i32,
}

// TODO: different representation?
pub enum EncodedGlyph {
    OneByte(u8),
    TwoBytes([u8; 2]),
}

pub trait Font {
    type Shaped<'a>: Iterator<Item = ShapedGlyph> + Clone + 'a
    where
        Self: 'a;

    fn shape<'a>(
        &'a self,
        text: &'a str,
        character_spacing: i32,
        word_spacing: i32,
    ) -> Self::Shaped<'a>;

    fn encode(&self, pdf: &mut crate::Pdf, glyph_id: u32, text: &str) -> EncodedGlyph;

    fn resource_name(&self) -> pdf_writer::Name<'_>;

    fn general_metrics(&self) -> GeneralMetrics;

    fn units_per_em(&self) -> u16;
}
