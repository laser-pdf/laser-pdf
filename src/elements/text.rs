use fonts::GeneralMetrics;
use text::{Line, Lines, draw_line, lines};
use utils::{mm_to_pt, pt_to_mm, set_fill_color};

use crate::{fonts::Font, *};

#[derive(Copy, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub enum TextAlign {
    Left,
    Center,
    Right,
}

/// A text element that renders text content with various styling options.
pub struct Text<'a, F: Font> {
    /// The text content to render
    pub text: &'a str,
    /// Font reference
    pub font: &'a F,
    /// Font size in points
    pub size: f32,
    /// Text color as RGBA (default: black 0x00_00_00_FF)
    pub color: u32,
    /// Whether to underline the text
    pub underline: bool,
    /// Additional spacing between characters
    pub extra_character_spacing: f32,
    /// Additional spacing between words
    pub extra_word_spacing: f32,
    /// Additional line height
    pub extra_line_height: f32,
    /// Text alignment
    pub align: TextAlign,
}

struct FontMetrics {
    ascent: f32,
    line_height: f32,
}

impl<'a, F: Font> Text<'a, F> {
    pub fn basic(text: &'a str, font: &'a F, size: f32) -> Self {
        Text {
            text,
            font,
            size,
            color: 0x00_00_00_FF,
            underline: false,
            extra_character_spacing: 0.,
            extra_word_spacing: 0.,
            extra_line_height: 0.,
            align: TextAlign::Left,
        }
    }

    fn compute_font_metrics(&self) -> FontMetrics {
        let GeneralMetrics {
            ascent,
            line_height,
        } = self.font.general_metrics();

        let units_per_em = self.font.units_per_em() as f32;

        FontMetrics {
            ascent: pt_to_mm(ascent as f32 * self.size / units_per_em),
            line_height: pt_to_mm(line_height as f32 * self.size / units_per_em)
                + self.extra_line_height,
        }
    }

    #[inline(always)]
    fn render_lines<'b, L: Iterator<Item = Line<F::Shaped<'a>>>>(
        &self,
        text: &str,
        lines: L,
        mut ctx: DrawCtx,
        ascent: f32,
        line_height: f32,
        width: f32,
    ) -> (f32, f32) {
        let mut max_width = width;
        let mut last_line_full_width = 0;

        let mut x = ctx.location.pos.0;
        let mut y = ctx.location.pos.1 - ascent;

        let mut height_available = ctx.first_height;

        let mut line_count = 0;
        let mut draw_rect = 0;

        let start = |pdf: &mut Pdf, location: &Location| {
            let layer = location.layer(pdf);
            layer
                .save_state()
                .set_font(self.font.resource_name(), self.size);
            set_fill_color(layer, self.color);
            layer.begin_text();
        };

        let end = |pdf: &mut Pdf, location: &Location| {
            location.layer(pdf).end_text().restore_state();
        };

        start(ctx.pdf, &ctx.location);

        for line in lines {
            let line_width =
                pt_to_mm(line.width as f32 / self.font.units_per_em() as f32 * self.size);
            max_width = max_width.max(line_width);

            last_line_full_width = line.width + line.trailing_whitespace_width;

            if height_available < line_height {
                if let Some(ref mut breakable) = ctx.breakable {
                    end(ctx.pdf, &ctx.location);

                    let new_location = (breakable.do_break)(
                        ctx.pdf,
                        draw_rect,
                        if line_count == 0 {
                            None
                        } else {
                            Some(line_count as f32 * line_height)
                        },
                    );
                    draw_rect += 1;
                    x = new_location.pos.0;
                    y = new_location.pos.1 - ascent;
                    height_available = breakable.full_height;
                    ctx.location.page_idx = new_location.page_idx;
                    ctx.location.layer_idx = new_location.layer_idx;
                    line_count = 0;

                    start(ctx.pdf, &ctx.location);
                }
            }

            let layer = ctx.location.layer(ctx.pdf);

            let x_offset = match self.align {
                TextAlign::Left => 0.,
                TextAlign::Center => (width - line_width) / 2.,
                TextAlign::Right => width - line_width,
            };

            let x = x + x_offset;

            layer.set_text_matrix([1.0, 0.0, 0.0, 1.0, mm_to_pt(x), mm_to_pt(y)]);

            draw_line(ctx.pdf, &ctx.location, self.font, text, line);

            y -= line_height;
            height_available -= line_height;
            line_count += 1;
        }

        end(ctx.pdf, &ctx.location);

        (
            max_width.max(pt_to_mm(
                last_line_full_width as f32 / self.font.units_per_em() as f32 * self.size,
            )),
            line_count as f32 * line_height,
        )
    }

    #[inline(always)]
    fn layout_lines<'b, L: Iterator<Item = Line<F::Shaped<'a>>>>(
        &self,
        lines: L,
        line_height: f32,
        measure_ctx: Option<&mut MeasureCtx>,
    ) -> (f32, f32) {
        let mut max_width = 0;
        let mut last_line_full_width = 0;
        let mut line_count = 0;

        // This function is a bit hacky because it's both used for measure and for determining the
        // max line width in unconstrained-width contexts.
        let mut height_available = if let Some(&mut MeasureCtx { first_height, .. }) = measure_ctx {
            first_height
        } else {
            f32::INFINITY
        };

        for line in lines {
            if let Some(&mut MeasureCtx {
                breakable: Some(ref mut breakable),
                ..
            }) = measure_ctx
            {
                if height_available < line_height {
                    *breakable.break_count += 1;
                    height_available = breakable.full_height;
                    line_count = 0;
                }
            }

            max_width = max_width.max(line.width);
            last_line_full_width = line.width + line.trailing_whitespace_width;

            height_available -= line_height;
            line_count += 1;
        }

        (
            pt_to_mm(
                max_width.max(last_line_full_width) as f32 / self.font.units_per_em() as f32
                    * self.size,
            ),
            line_count as f32 * line_height,
        )
    }

    fn break_into_lines<R>(
        &'a self,
        width: f32,
        f: impl for<'b> FnOnce(Lines<'a, 'b, F>) -> R,
    ) -> R {
        lines(
            self.font,
            (self.extra_character_spacing / self.size * self.font.units_per_em() as f32) as i32,
            (self.extra_word_spacing / self.size * self.font.units_per_em() as f32) as i32,
            // The ceil is to prevent rounding errors from causing problems in cases where the
            // element gets measured and then the measured width gets used for draw, such as in
            // HAlign.
            (mm_to_pt(width) / self.size * self.font.units_per_em() as f32).ceil() as u32,
            self.text,
            f,
        )
    }
}

impl<'a, F: Font> Element for Text<'a, F> {
    fn first_location_usage(&self, ctx: FirstLocationUsageCtx) -> FirstLocationUsage {
        let FontMetrics {
            ascent: _,
            line_height,
        } = self.compute_font_metrics();

        if line_height > ctx.first_height {
            FirstLocationUsage::WillSkip
        } else {
            FirstLocationUsage::WillUse
        }
    }

    fn measure(&self, mut ctx: MeasureCtx) -> ElementSize {
        let FontMetrics { line_height, .. } = self.compute_font_metrics();

        let size = self.break_into_lines(ctx.width.max, |lines| {
            self.layout_lines(lines, line_height, Some(&mut ctx))
        });

        ElementSize {
            width: Some(ctx.width.constrain(size.0)),
            height: Some(size.1),
        }
    }

    fn draw(&self, ctx: DrawCtx) -> ElementSize {
        let FontMetrics {
            ascent,
            line_height,
        } = self.compute_font_metrics();

        // For left alignment we don't need to pre-layout because the
        // x offset is always zero.
        let width = if ctx.width.expand {
            ctx.width.max
        } else if self.align == TextAlign::Left {
            0.
        } else {
            // TODO: Figure out a way to avoid shaping twice here.
            self.break_into_lines(ctx.width.max, |lines| {
                self.layout_lines(lines, line_height, None).0
            })
        };

        let width_constraint = ctx.width;
        let size = self.break_into_lines(ctx.width.max, |lines| {
            self.render_lines(self.text, lines, ctx, ascent, line_height, width)
        });

        ElementSize {
            width: Some(width_constraint.constrain(size.0)),
            height: Some(size.1),
        }
    }
}

#[cfg(test)]
mod tests {
    use elements::column::Column;
    use fonts::truetype::TruetypeFont;
    use insta::*;

    use crate::test_utils::binary_snapshots::*;
    use crate::{
        DrawCtx, ElementSize,
        fonts::builtin::BuiltinFont,
        test_utils::{ElementProxy, ElementTestParams},
    };

    use super::*;

    const FONT: &[u8] = include_bytes!("../fonts/Kenney Bold.ttf");

    #[test]
    fn test_multi_page() {
        let bytes = test_element_bytes(TestElementParams::breakable(), |mut callback| {
            let font = BuiltinFont::courier(callback.pdf());

            let content = Text::basic(LOREM_IPSUM, &font, 32.);
            let content = content.debug(0);

            callback.call(&content);
        });
        assert_binary_snapshot!(".pdf", bytes);
    }

    #[test]
    fn test_truetype() {
        let bytes = test_element_bytes(TestElementParams::breakable(), |mut callback| {
            let font = TruetypeFont::new(callback.pdf(), FONT);

            let content = Text::basic(LOREM_IPSUM, &font, 32.);
            let content = content.debug(0);

            callback.call(&content);
        });
        assert_binary_snapshot!(".pdf", bytes);
    }

    #[test]
    fn test_truetype_trailing_whitespace() {
        let mut params = TestElementParams::breakable();
        params.width.expand = false;

        let bytes = test_element_bytes(params, |mut callback| {
            let font = TruetypeFont::new(callback.pdf(), FONT);

            let content = Text::basic("Whitespace ", &font, 32.);
            let content = content.debug(0);

            callback.call(&content);
        });
        assert_binary_snapshot!(".pdf", bytes);
    }

    #[test]
    fn test_truetype_extra_spacing() {
        let mut params = TestElementParams::breakable();
        params.width.expand = false;

        let bytes = test_element_bytes(params, |mut callback| {
            let font = TruetypeFont::new(callback.pdf(), FONT);

            callback.call(&Column {
                gap: 12.,
                collapse: false,
                content: |content| {
                    let normal = Text::basic("Hello, World", &font, 32.);

                    let character_spacing = Text {
                        extra_character_spacing: 16.,
                        ..Text::basic("Hello, World", &font, 32.)
                    };

                    let word_spacing = Text {
                        extra_word_spacing: 16.,
                        ..Text::basic("Hello, World", &font, 32.)
                    };

                    let both = Text {
                        extra_character_spacing: 16.,
                        extra_word_spacing: 16.,
                        ..Text::basic("Hello, World", &font, 32.)
                    };

                    content
                        .add(&normal.debug(0).show_max_width())?
                        .add(&character_spacing.debug(1).show_max_width())?
                        .add(&word_spacing.debug(2).show_max_width())?
                        .add(&both.debug(3).show_max_width())?;

                    None
                },
            });
        });
        assert_binary_snapshot!(".pdf", bytes);
    }

    #[test]
    fn test_truetype_soft_hyphen() {
        let mut params = TestElementParams::breakable();
        params.width.expand = false;

        let bytes = test_element_bytes(params, |mut callback| {
            let font = TruetypeFont::new(callback.pdf(), FONT);

            callback.call(&Column {
                gap: 12.,
                collapse: false,
                content: |content| {
                    let a = Text::basic("Hello\u{00AD}Wrld", &font, 32.);
                    let b = Text::basic("A Hello\u{00AD}Wrld", &font, 32.);
                    let c = Text::basic("A\u{00A0}Hello\u{00AD}Wrld", &font, 32.);
                    let d = Text::basic("Hello\u{00AD}Wrld\u{00AD}", &font, 32.);

                    content
                        .add(&Padding::right(100., a.debug(0).show_max_width()))?
                        .add(&Padding::right(120., b.debug(0).show_max_width()))?
                        .add(&Padding::right(120., c.debug(0).show_max_width()))?
                        .add(&Padding::right(20., d.debug(0).show_max_width()))?;

                    None
                },
            });
        });
        assert_binary_snapshot!(".pdf", bytes);
    }

    // #[test]
    // fn test_text() {
    //     // A fake document for adding the font to.
    //     let doc = PdfDocument::empty("i contain a font");

    //     let font = BuiltinFont::helvetica(&doc);

    //     let text_element = Text {
    //         ..Text::basic("i am a line\nso am i", &font, 12.)
    //     };

    //     let element = ElementProxy {
    //         before_draw: &|ctx: &mut DrawCtx| {
    //             // These seem to be stored in a map by name and when drawing a font with the same
    //             // needs to exist in the document being drawn on.
    //             ctx.pdf
    //                 .document
    //                 .add_builtin_font(printpdf::BuiltinFont::Helvetica)
    //                 .unwrap();
    //         },
    //         ..ElementProxy::new(text_element)
    //     };

    //     for mut output in (ElementTestParams {
    //         first_height: 4.,
    //         full_height: 5.,
    //         ..Default::default()
    //     })
    //     .run(&element)
    //     {
    //         if let Some(ref mut b) = output.breakable {
    //             b.assert_break_count(if output.first_height == 4. { 2 } else { 1 });
    //         }

    //         output.assert_size(ElementSize {
    //             width: Some(output.width.constrain(19.291312152)),

    //             // Note: I'm not sure this line height is correct. When running the same test with
    //             // Nimbus Sans L, which is supposed to be fully metrically compatible with
    //             // helvetica (at least according to the readme in
    //             // https://git.ghostscript.com/?p=user/tor/urw-base-12.git), the height ends up
    //             // being slightly more. On the x axis it matches exactly though. It's possible that
    //             // the bounding box in afm is not meant to be the equivalent of ascent + descent +
    //             // line gap in ttf. NimbusSans-Regular.afm from the following repo has the same
    //             // numbers https://git.ghostscript.com/?p=urw-core35-fonts.git as the Adobe one
    //             // and when running with NimbusSans-Regular.ttf from that repo the numbers are much
    //             // closer, which is reassuring (that one uses 2048 units per em so that should
    //             // explain the slight difference). The numbers for that one are 19.293924914062497
    //             // and 4.86792298828125.
    //             height: Some(4.893736415999999 * if output.breakable.is_some() { 1. } else { 2. }),
    //         });
    //     }
    // }
}
