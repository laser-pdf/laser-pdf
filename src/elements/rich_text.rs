use crate::fonts::Font;
use crate::fonts::GeneralMetrics;
// use crate::text::remove_non_trailing_soft_hyphens;
use crate::text::*;
use crate::utils::*;
use crate::*;

use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct Span {
    pub text: String,
    pub bold: bool,
    pub italic: bool,
    pub underline: bool,
    pub color: u32,
}

pub struct RichText<'a, F: Font> {
    pub spans: &'a [Span],
    pub size: f32,
    pub small_size: f32,
    pub extra_line_height: f32,
    pub fonts: FontSet<'a, F>,
}

impl<'a, F: Font> Element for RichText<'a, F> {
    fn first_location_usage(&self, ctx: FirstLocationUsageCtx) -> FirstLocationUsage {
        let line_height = self.line_height();
        let line_height = line_height + self.extra_line_height;

        if ctx.first_height < line_height {
            FirstLocationUsage::WillSkip
        } else {
            FirstLocationUsage::WillUse
        }
    }

    fn measure(&self, mut ctx: MeasureCtx) -> ElementSize {
        let mut max_width = ctx.width.constrain(0.);

        let line_height = self.line_height();
        let line_height = line_height + self.extra_line_height;

        let mut height_available = ctx.first_height;

        if height_available < line_height {
            if let Some(ref mut breakable) = ctx.breakable {
                *breakable.break_count += 1;
                height_available = breakable.full_height;
            }
        }

        let mut line_count = 1;

        self.line_fragments(ctx.width.max, |frag| {
            let line_width = frag.length;

            max_width = max_width.max(frag.x_offset + line_width);

            if frag.new_line {
                match ctx.breakable {
                    Some(ref mut breakable) if height_available < 2. * line_height => {
                        *breakable.break_count += 1;

                        height_available = breakable.full_height;
                        line_count = 1;
                    }
                    _ => {
                        height_available -= line_height;
                        line_count += 1;
                    }
                }
            }
        });

        ElementSize {
            width: Some(max_width),
            height: Some(line_count as f32 * line_height),
        }
    }

    fn draw(&self, mut ctx: DrawCtx) -> ElementSize {
        let mut max_width = ctx.width.constrain(0.);

        let line_height = self.line_height();
        let line_height = line_height + self.extra_line_height;

        let mut x = ctx.location.pos.0;
        let mut y = ctx.location.pos.1;

        let mut height_available = ctx.first_height;

        let mut draw_rect = 0;

        if height_available < line_height {
            if let Some(ref mut breakable) = ctx.breakable {
                let new_location = (breakable.do_break)(ctx.pdf, draw_rect, None);
                draw_rect = 1;
                x = new_location.pos.0;
                y = new_location.pos.1;
                height_available = breakable.full_height;
                ctx.location.page_idx = new_location.page_idx;
                ctx.location.layer_idx = new_location.layer_idx;
            }
        }

        let mut line_count = 1;

        self.line_fragments(ctx.width.max, |frag| {
            let line_width = frag.length;

            max_width = max_width.max(frag.x_offset + line_width);

            if frag.new_line {
                match ctx.breakable {
                    Some(ref mut breakable) if height_available < 2. * line_height => {
                        let new_location = (breakable.do_break)(
                            ctx.pdf,
                            draw_rect,
                            Some(line_count as f32 * line_height),
                        );
                        draw_rect += 1;

                        x = new_location.pos.0;
                        y = new_location.pos.1;
                        height_available = breakable.full_height;
                        ctx.location.page_idx = new_location.page_idx;
                        ctx.location.layer_idx = new_location.layer_idx;
                        line_count = 1;
                    }
                    _ => {
                        y -= line_height;
                        height_available -= line_height;
                        line_count += 1;
                    }
                }
            }

            let layer = ctx.location.layer(ctx.pdf);
            layer.save_state();

            set_fill_color(layer, frag.span.color);

            let layer = ctx.location.layer(ctx.pdf);

            layer.save_state();
            set_fill_color(layer, frag.span.color);

            layer
                .set_font(frag.font.resource_name(), self.size)
                .set_text_matrix([
                    1.0,
                    0.0,
                    0.0,
                    1.0,
                    mm_to_pt(x + frag.x_offset),
                    mm_to_pt(y - line_height + frag.font_vars.line_height - frag.font_vars.ascent),
                ]);

            draw_line(
                ctx.pdf,
                &ctx.location,
                frag.font,
                &frag.span.text,
                frag.line,
            );

            ctx.location.layer(ctx.pdf).restore_state();
        });

        ElementSize {
            width: Some(max_width),
            height: Some(line_count as f32 * line_height),
        }
    }
}

struct LineFragment<'a, F: Font> {
    font: &'a F,
    font_vars: &'a FontVars,

    span: &'a Span,
    line: Line<F::Shaped<'a>>,

    /// Whether the fragment goes on a new line. So the line breaking has to happen before the
    /// fragment, not after.
    new_line: bool,

    length: f32,

    x_offset: f32,
}

impl<'a, F: Font> RichText<'a, F> {
    // Currently has to be an internal iterator because of LineGenerator, which works with a
    // callback. We'll probably have to change this at some point if we want to support justified
    // text. But for that we'll also need to solve the problem of the unicode segmentation iterator
    // not being cloneable.
    fn line_fragments(&self, width: f32, mut f: impl for<'b> FnMut(LineFragment<'b, F>)) {
        let regular_vars = font_vars(self.fonts.regular, self.size as f32);
        let bold_vars = font_vars(self.fonts.bold, self.size as f32);
        let italic_vars = font_vars(self.fonts.italic, self.size as f32);
        let bold_italic_vars = font_vars(self.fonts.bold_italic, self.size as f32);

        #[derive(PartialEq, Eq, Debug)]
        enum LineState {
            FirstLine,
            InLine,
            LineDone,
        }

        use LineState::*;

        let mut line_state = FirstLine;

        let mut x_offset_pt = 0.;

        for span in self.spans {
            let (font, font_vars): (&F, FontVars) = match (span.bold, span.italic) {
                (false, false) => (self.fonts.regular, regular_vars),
                (false, true) => (self.fonts.italic, italic_vars),
                (true, false) => (self.fonts.bold, bold_vars),
                (true, true) => (self.fonts.bold_italic, bold_italic_vars),
            };

            let mut last_line_empty = true;

            LineGenerator::new(font, &span.text, |mut generator| {
                while let Some(line) = generator.next(
                    ((mm_to_pt(width)
                        - (line_state == InLine).then_some(x_offset_pt).unwrap_or(0.))
                        / self.size
                        * font.units_per_em() as f32) as u32,
                    line_state == InLine,
                ) {
                    if line_state != InLine {
                        x_offset_pt = 0.;
                    }

                    let length =
                        pt_to_mm(line.width as f32 / font.units_per_em() as f32 * self.size);

                    let width = line.width;
                    let trailing_whitespace_width = line.trailing_whitespace_width;

                    last_line_empty = last_line_empty && line.empty;

                    // We need empty parts at the beginning of lines, otherwise trailing newlines
                    // on spans don't work. The reason we filter out empty fragments at all is so
                    // that we don't need add trailing whitespace to the width.
                    if !line.empty || line_state != InLine {
                        f(LineFragment {
                            font,
                            font_vars: &font_vars,
                            span,
                            line,
                            new_line: line_state == LineDone,
                            x_offset: pt_to_mm(x_offset_pt),
                            length,
                        });
                    }

                    x_offset_pt += (width + trailing_whitespace_width) as f32
                        / font.units_per_em() as f32
                        * self.size;
                    line_state = LineDone;
                }
            });

            line_state = if last_line_empty { FirstLine } else { InLine };
        }
    }

    fn line_height(&self) -> f32 {
        let regular_vars = font_vars(self.fonts.regular, self.size as f32);
        let bold_vars = font_vars(self.fonts.bold, self.size as f32);
        let italic_vars = font_vars(self.fonts.italic, self.size as f32);
        let bold_italic_vars = font_vars(self.fonts.bold_italic, self.size as f32);

        regular_vars
            .line_height
            .max(bold_vars.line_height)
            .max(italic_vars.line_height)
            .max(bold_italic_vars.line_height)
    }
}

#[derive(Copy, Clone)]
struct FontVars {
    ascent: f32,
    line_height: f32,
}

fn font_vars<F: Font>(font: &F, size: f32) -> FontVars {
    let GeneralMetrics {
        ascent,
        line_height,
    } = font.general_metrics();

    let units_per_em = font.units_per_em() as f32;

    FontVars {
        ascent: pt_to_mm(ascent as f32 * size / units_per_em),
        line_height: pt_to_mm(line_height as f32 * size / units_per_em),
    }
}

#[cfg(test)]
mod tests {
    use elements::column::{Column, ColumnContent};
    use fonts::{ShapedGlyph, truetype::TruetypeFont};
    use insta::*;

    use crate::test_utils::binary_snapshots::*;

    use super::*;

    #[derive(Debug)]
    struct FakeFont;

    #[derive(Clone, Debug)]
    struct FakeShaped<'a> {
        // last: usize,
        inner: std::str::CharIndices<'a>,
    }

    impl<'a> Iterator for FakeShaped<'a> {
        type Item = ShapedGlyph;

        fn next(&mut self) -> Option<Self::Item> {
            if let Some((i, c)) = self.inner.next() {
                Some(ShapedGlyph {
                    unsafe_to_break: false,
                    glyph_id: c as u32,
                    text_range: i..i + c.len_utf8(),
                    x_advance: if matches!(c, '\u{00ad}') { 0 } else { 1 },
                    x_offset: 0,
                    y_offset: 0,
                    y_advance: 0,
                })
            } else {
                None
            }
        }
    }

    impl Font for FakeFont {
        type Shaped<'a>
            = FakeShaped<'a>
        where
            Self: 'a;

        fn shape<'a>(&'a self, text: &'a str) -> Self::Shaped<'a> {
            FakeShaped {
                inner: text.char_indices(),
            }
        }

        fn encode(&self, _: &mut crate::Pdf, _: u32, _: &str) -> crate::fonts::EncodedGlyph {
            unimplemented!()
        }

        fn resource_name(&self) -> pdf_writer::Name {
            unimplemented!()
        }

        fn general_metrics(&self) -> GeneralMetrics {
            GeneralMetrics {
                ascent: 0,
                line_height: 1,
            }
        }

        fn units_per_em(&self) -> u16 {
            1
        }
    }

    #[test]
    fn test_line_fragments() {
        let font = FakeFont;

        let element = RichText {
            spans: &[
                Span {
                    text: "Where are ".to_string(),
                    bold: false,
                    italic: false,
                    underline: false,
                    color: 0x00_00_00_FF,
                },
                Span {
                    text: "they".to_string(),
                    bold: true,
                    italic: false,
                    underline: false,
                    color: 0x00_00_FF_FF,
                },
                Span {
                    text: "\n".to_string(),
                    bold: false,
                    italic: false,
                    underline: false,
                    color: 0x00_00_FF_FF,
                },
                Span {
                    text: "at?".to_string(),
                    bold: false,
                    italic: false,
                    underline: false,
                    color: 0xFF_00_00_FF,
                },
            ],
            size: 1.,
            small_size: 1.,
            extra_line_height: 0.,
            fonts: FontSet {
                regular: &font,
                bold: &font,
                italic: &font,
                bold_italic: &font,
            },
        };

        #[derive(Debug)]
        #[allow(dead_code)]
        struct Frag {
            new_line: bool,
            text: String,
            x_offset: f32,
            span: usize,
            empty: bool,
        }

        let collect = |width: f32| -> Vec<Frag> {
            let mut results = Vec::new();

            element.line_fragments(width, |mut fragment| {
                let empty = fragment.line.empty;
                let text = if let Some(first) = fragment.line.next() {
                    let last = fragment.line.last();

                    &fragment.span.text[first.text_range.start
                        ..last.map_or(first.text_range.end, |l| l.text_range.end)]
                } else {
                    ""
                };

                results.push(Frag {
                    text: text.to_string(),
                    new_line: fragment.new_line,
                    x_offset: mm_to_pt(fragment.x_offset),
                    span: element
                        .spans
                        .iter()
                        .enumerate()
                        .find_map(|(i, span)| std::ptr::eq(span, fragment.span).then_some(i))
                        .unwrap(),
                    empty,
                });
            });

            results
        };

        assert_debug_snapshot!(collect(0.), @r#"
        [
            Frag {
                new_line: false,
                text: "Where ",
                x_offset: 0.0,
                span: 0,
                empty: false,
            },
            Frag {
                new_line: true,
                text: "are ",
                x_offset: 0.0,
                span: 0,
                empty: false,
            },
            Frag {
                new_line: true,
                text: "they",
                x_offset: 0.0,
                span: 1,
                empty: false,
            },
            Frag {
                new_line: true,
                text: "",
                x_offset: 0.0,
                span: 2,
                empty: true,
            },
            Frag {
                new_line: false,
                text: "at?",
                x_offset: 0.0,
                span: 3,
                empty: false,
            },
        ]
        "#);
        assert_debug_snapshot!(collect(3.), @r#"
        [
            Frag {
                new_line: false,
                text: "Where ",
                x_offset: 0.0,
                span: 0,
                empty: false,
            },
            Frag {
                new_line: true,
                text: "are ",
                x_offset: 0.0,
                span: 0,
                empty: false,
            },
            Frag {
                new_line: false,
                text: "they",
                x_offset: 4.0,
                span: 1,
                empty: false,
            },
            Frag {
                new_line: true,
                text: "",
                x_offset: 0.0,
                span: 2,
                empty: true,
            },
            Frag {
                new_line: false,
                text: "at?",
                x_offset: 0.0,
                span: 3,
                empty: false,
            },
        ]
        "#);
        assert_debug_snapshot!(collect(13.), @r#"
        [
            Frag {
                new_line: false,
                text: "Where are ",
                x_offset: 0.0,
                span: 0,
                empty: false,
            },
            Frag {
                new_line: false,
                text: "they",
                x_offset: 10.0,
                span: 1,
                empty: false,
            },
            Frag {
                new_line: true,
                text: "",
                x_offset: 0.0,
                span: 2,
                empty: true,
            },
            Frag {
                new_line: false,
                text: "at?",
                x_offset: 0.0,
                span: 3,
                empty: false,
            },
        ]
        "#);
    }

    #[test]
    fn test_truetype() {
        let bytes = test_element_bytes(TestElementParams::breakable(), |mut callback| {
            let regular =
                TruetypeFont::new(callback.pdf(), include_bytes!("../fonts/Kenney Future.ttf"));
            let bold =
                TruetypeFont::new(callback.pdf(), include_bytes!("../fonts/Kenney Bold.ttf"));

            let rich_text = RichText {
                spans: &[
                    Span {
                        text: "Where are ".to_string(),
                        bold: false,
                        italic: false,
                        underline: false,
                        color: 0x00_00_00_FF,
                    },
                    Span {
                        text: "they".to_string(),
                        bold: true,
                        italic: false,
                        underline: false,
                        color: 0x00_00_FF_FF,
                    },
                    Span {
                        text: "\n".to_string(),
                        bold: true,
                        italic: false,
                        underline: false,
                        color: 0x00_00_FF_FF,
                    },
                    Span {
                        text: "at?".to_string(),
                        bold: false,
                        italic: false,
                        underline: false,
                        color: 0xFF_00_00_FF,
                    },
                ],
                size: 12.,
                small_size: 8.,
                extra_line_height: 0.,
                fonts: FontSet {
                    regular: &regular,
                    bold: &bold,
                    italic: &regular,
                    bold_italic: &bold,
                },
            };

            let list = Column {
                gap: 16.,
                collapse: false,
                content: |content: ColumnContent| {
                    content
                        .add(&rich_text.debug(0))?
                        .add(&Padding::right(140., &rich_text.debug(1).show_max_width()))?
                        .add(&Padding::right(160., &rich_text.debug(2).show_max_width()))?
                        .add(&Padding::right(180., &rich_text.debug(3).show_max_width()))?
                        .add(&Padding::right(194., &rich_text.debug(4).show_max_width()))?;
                    None
                },
            };

            callback.call(&list);
        });
        assert_binary_snapshot!(".pdf", bytes);
    }
}
