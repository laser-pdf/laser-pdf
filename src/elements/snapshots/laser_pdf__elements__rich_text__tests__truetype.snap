---
source: src/elements/rich_text.rs
expression: bytes
extension: pdf
snapshot_kind: binary
---
